File: scripts/run_optimization.py
# ══════════════════════════════════════════════════════════════════
# RUN_OPTIMIZATION
# Top-level entry point: load config, run the full optimization pipeline,
# save weights, metrics, and plots to a timestamped results folder.
#
# Part of: Antenna Array Pattern Optimization Tool
# ══════════════════════════════════════════════════════════════════

# ────────────────────────── IMPORTS ───────────────────────────────
import argparse
import csv
import json
import shutil
import sys
import time
from datetime import datetime
from pathlib import Path

# Ensure the project root (parent of scripts/) is on sys.path so that
# `src.*` imports resolve when the script is invoked as a file rather
# than as a module (i.e. `python scripts/run_optimization.py`).
sys.path.insert(0, str(Path(__file__).parent.parent))
# [MATLAB] addpath(fullfile(fileparts(mfilename('fullpath')), '..'));

import numpy as np
import yaml

from src.io.cst_parser import load_element_patterns
from src.cost.cost_function import compute_array_factor
from src.optimize.optimizer import run_optimizer
from src.metrics.metrics import evaluate_metrics
from src.plot.plotter import save_all_plots, save_pattern_gif

# ────────────────────────── CONSTANTS ─────────────────────────────

# Polarization modes wired through the optimizer pipeline.
# "copol" → E_complex (co-pol), "cross" → cross_complex (cross-pol).
# "total" is supported only in the interactive manual_weights tool, where the
# power-sum |AF_copol|² + |AF_xpol|² is rendered without a single coherent AF.
SUPPORTED_POLARIZATIONS = ("copol", "cross")

# Mapping from polarization name → pattern dict key used to build the stack.
POLARIZATION_TO_PATTERN_KEY = {
    "copol": "E_complex",
    "cross": "cross_complex",
}

# Required top-level keys in config.yaml.
REQUIRED_CONFIG_KEYS = ("element_patterns_dir", "directives", "optimizer", "output")


# ────────────────────────── HELPERS ───────────────────────────────

def _load_config(config_path):
    """Load config.yaml and return its contents as a dict.

    Args:
        config_path (str | Path): Path to the YAML configuration file.

    Returns:
        dict: Parsed configuration dictionary.

    Raises:
        FileNotFoundError: If the config file does not exist at config_path.
    """
    with open(config_path, "r", encoding="utf-8") as fh:
        return yaml.safe_load(fh)
    # [MATLAB] config = yaml_load(config_path);  % requires a YAML toolbox


def _validate_config(config):
    """Check that all required config keys are present and polarization is supported.

    Args:
        config (dict): Parsed configuration dictionary.

    Raises:
        KeyError: If a required top-level key is absent.
        ValueError: If the specified polarization is not yet implemented.
    """
    for key in REQUIRED_CONFIG_KEYS:
        if key not in config:
            raise KeyError(
                f"Missing required config key: '{key}'. "
                f"Required keys: {REQUIRED_CONFIG_KEYS}."
            )

    polarization = config.get("polarization", "copol")
    if polarization not in SUPPORTED_POLARIZATIONS:
        raise ValueError(
            f"Polarization '{polarization}' is not yet implemented. "
            f"Supported options: {SUPPORTED_POLARIZATIONS}. "
            "Set polarization: \"copol\" in config.yaml."
        )


def _create_output_dir(results_dir):
    """Create and return a timestamped output subfolder under results_dir.

    Args:
        results_dir (str | Path): Base results directory from config.

    Returns:
        Path: Newly created timestamped output directory.
    """
    timestamp  = datetime.now().strftime("%Y-%m-%d_%H%M%S")
    # [MATLAB] timestamp = datestr(now, 'yyyy-mm-dd_HHMMSS');
    output_dir = Path(results_dir) / timestamp
    output_dir.mkdir(parents=True, exist_ok=True)
    return output_dir


def _save_weights_csv(weights_complex, output_path):
    """Save optimized complex weights to a CSV file.

    Columns: element_index, amplitude, phase_deg, real, imag.

    Args:
        weights_complex (np.ndarray): Optimized complex weights, shape (N_elements,).
            Units: dimensionless (V/V).
        output_path (str | Path): Full path for the CSV output file.
    """
    with open(output_path, "w", newline="") as fh:
        writer = csv.writer(fh)
        writer.writerow(["element_index", "amplitude", "phase_deg", "real", "imag"])
        for n, w in enumerate(weights_complex):
            writer.writerow([
            n,
            f"{np.abs(w):.2f}",
            f"{np.angle(w, deg=True):.2f}",
            f"{w.real:.2f}",
            f"{w.imag:.2f}"
        ])
    # [MATLAB] writematrix([indices, amps, phases, reals, imags], output_path);


def _save_metrics_json(metrics, output_path):
    """Save the metrics dict to a JSON file with 2-space indentation.

    All values in the metrics dict are native Python types (float, int, str, None),
    so no custom JSON encoder is required.

    Args:
        metrics (dict): Output of evaluate_metrics(...).
        output_path (str | Path): Full path for the JSON output file.
    """
    with open(output_path, "w") as fh:
        json.dump(metrics, fh, indent=2)
    # [MATLAB] jsonencode(metrics) — single-line only; no indent support


def _save_run_report(metrics, optimizer_result, elapsed_sec, config, output_path):
    """Save a human-readable run report for cross-run comparison.

    Args:
        metrics (dict): Output of evaluate_metrics(...).
        optimizer_result (dict): Output of run_optimizer(...).
        elapsed_sec (float): Wall-clock optimization time in seconds.
        config (dict): Full parsed configuration dictionary.
        output_path (str | Path): Full path for the text output file.
    """
    converged     = optimizer_result["result"].success
    opt_cfg       = config.get("optimizer", {})
    directives    = config.get("directives", [])
    timestamp_str = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

    lines = [
        "=" * 56,
        "Antenna Array Optimization Report",
        timestamp_str,
        "=" * 56,
        "",
        "RUN INFO",
        f"  Elapsed time:      {elapsed_sec:.1f} s",
        f"  Converged:         {converged}",
        f"  Iterations:        {metrics['iteration_count']}",
        f"  Final cost J:      {metrics['total_cost']:.6f}",
        "",
        "OPTIMIZER SETTINGS",
        "  Method:            L-BFGS-B",
        f"  Max iterations:    {opt_cfg.get('max_iterations')}",
        f"  Cost tolerance:    {opt_cfg.get('cost_tolerance')}",
        f"  N restarts:        {opt_cfg.get('n_restarts')}",
        f"  Amplitude bounds:  {opt_cfg.get('amplitude_bounds')}",
        f"  Phase only:        {opt_cfg.get('phase_only', False)}",
        f"  Amplitude only:    {opt_cfg.get('amplitude_only', False)}",
        "",
        "DIRECTIVES",
    ]
    for i, d in enumerate(directives):
        lines.append(
            f"  [{i}] {d['type']:<4}  theta={d['theta']:.1f}°  "
            f"phi={d.get('phi', 0.0):.1f}°  "
            f"width={d['width']:.1f}°  weight={d.get('weight', 1.0)}"
        )

    lines += [
        "",
        "RESULTS",
        f"  Global peak:       {metrics['global_peak_dbi']:.2f} dBi",
    ]
    if metrics["peak_to_null_ratio_db"] is not None:
        lines.append(f"  Peak-to-null:      {metrics['peak_to_null_ratio_db']:.2f} dB")
    lines.append("")

    for dm in metrics["directive_metrics"]:
        if dm["type"] == "peak":
            lines.append(
                f"  Peak @ theta={dm['theta_deg']:.1f}°, phi={dm['phi_deg']:.1f}°:  "
                f"{dm['gain_dbi']:.2f} dBi"
            )
        else:
            lines.append(
                f"  Null @ theta={dm['theta_deg']:.1f}°, phi={dm['phi_deg']:.1f}°:  "
                f"{dm['gain_dbi']:.2f} dBi  (depth = {dm['null_depth_db']:.2f} dB)"
            )
    lines += ["", "=" * 56]

    with open(output_path, "w", encoding="utf-8") as fh:
        fh.write("\n".join(lines) + "\n")


def _print_summary(metrics, optimizer_result):
    """Print a human-readable optimization summary to stdout.

    Args:
        metrics (dict): Output of evaluate_metrics(...).
        optimizer_result (dict): Output of run_optimizer(...).
    """
    converged = optimizer_result["result"].success
    print(f"  Converged:          {converged}")
    print(f"  Iterations:         {metrics['iteration_count']}")
    print(f"  Final cost J:       {metrics['total_cost']:.6f}")
    print(f"  Global peak:        {metrics['global_peak_dbi']:.2f} dBi")

    for dm in metrics["directive_metrics"]:
        if dm["type"] == "peak":
            print(f"  Peak @ theta={dm['theta_deg']}°, phi={dm['phi_deg']}°:  "
                  f"{dm['gain_dbi']:.2f} dBi")
        else:
            print(f"  Null @ theta={dm['theta_deg']}°, phi={dm['phi_deg']}°:  "
                  f"{dm['gain_dbi']:.2f} dBi  "
                  f"(depth = {dm['null_depth_db']:.2f} dB)")

    if metrics["peak_to_null_ratio_db"] is not None:
        print(f"  Peak-to-null:       {metrics['peak_to_null_ratio_db']:.2f} dB")


# ────────────────────────── MAIN ──────────────────────────────────

def main():
    """Run the full antenna array pattern optimization pipeline."""
    # ── Parse arguments ────────────────────────────────────────────
    arg_parser = argparse.ArgumentParser(
        description="Antenna Array Pattern Optimization"
    )
    arg_parser.add_argument(
        "--config", default="config.yaml",
        help="Path to the YAML configuration file (default: config.yaml).",
    )
    args = arg_parser.parse_args()

    # ── Load and validate config ───────────────────────────────────
    config = _load_config(args.config)
    _validate_config(config)

    # ── Create timestamped output directory ────────────────────────
    output_dir = _create_output_dir(config["output"]["results_dir"])
    print(f"Output directory: {output_dir}")

    # ── Stage 1: Load element patterns ────────────────────────────
    print("Loading element patterns...")
    patterns = load_element_patterns(config["element_patterns_dir"])

    # Pick the per-element complex pattern according to the chosen polarization.
    polarization = config.get("polarization", "copol")
    pattern_key  = POLARIZATION_TO_PATTERN_KEY[polarization]
    print(f"  Polarization: {polarization} (using '{pattern_key}')")

    # Stack individual element patterns into a single 3D array.
    element_patterns_stacked = np.stack([p[pattern_key] for p in patterns], axis=0)
    # [MATLAB] element_patterns_stacked = cat(3, patterns{:}.(pattern_key));
    theta_deg = patterns[0]["theta_deg"]
    phi_deg   = patterns[0]["phi_deg"]

    n_elements, n_theta, n_phi = element_patterns_stacked.shape
    print(f"  Loaded {n_elements} elements, grid {n_theta}×{n_phi}")

    # ── Stage 4: Run optimizer ─────────────────────────────────────
    print("Running optimizer...")
    directives    = config["directives"]
    t_start       = time.perf_counter()
    optimizer_result = run_optimizer(
        element_patterns_stacked, theta_deg, phi_deg,
        directives, config["optimizer"],
    )
    elapsed_sec     = time.perf_counter() - t_start
    weights_complex    = optimizer_result["weights_complex"]
    cost_history       = optimizer_result["cost_history"]
    all_cost_histories = optimizer_result["all_cost_histories"]
    best_run_index     = optimizer_result["best_run_index"]
    all_run_labels     = optimizer_result["all_run_labels"]

    # ── Stage 5: Evaluate metrics ──────────────────────────────────
    metrics = evaluate_metrics(
        element_patterns_stacked, theta_deg, phi_deg,
        weights_complex, directives, cost_history,
    )

    # ── Compute array factor dB grid for plotting ──────────────────
    # Array factor: coherent superposition of weighted element patterns.
    array_factor = compute_array_factor(weights_complex, element_patterns_stacked)
    power_linear = np.abs(array_factor) ** 2
    # Guard against log(0) before converting to dB.
    array_factor_db_grid = 10.0 * np.log10(np.maximum(power_linear, 1e-30))
    # [MATLAB] array_factor_db_grid = 10 * log10(max(abs(array_factor).^2, 1e-30));

    # ── Stage 6: Save plots ────────────────────────────────────────
    print("Saving plots...")
    save_all_plots(
        array_factor_db_grid, theta_deg, phi_deg,
        weights_complex, directives,
        all_cost_histories, best_run_index, all_run_labels,
        config["output"], output_dir,
    )

    # ── Stage 7: Save pattern evolution GIF (optional) ────────────
    if config["output"].get("save_pattern_gif", False):
        print("Saving pattern evolution GIF...")
        save_pattern_gif(
            optimizer_result["weights_history"],
            element_patterns_stacked,
            theta_deg, phi_deg, directives,
            cost_history, config["output"], output_dir,
        )

    # ── Save weights and metrics ───────────────────────────────────
    _save_weights_csv(weights_complex, output_dir / "weights.csv")
    _save_metrics_json(metrics,        output_dir / "metrics.json")

    # ── Save run report and config snapshot ───────────────────────
    _save_run_report(metrics, optimizer_result, elapsed_sec, config,
                     output_dir / "run_report.txt")
    shutil.copy(args.config, output_dir / "config.yaml")

    # ── Print summary ──────────────────────────────────────────────
    print("\nResults:")
    print(f"  Optimization time:  {elapsed_sec:.1f} s")
    _print_summary(metrics, optimizer_result)
    print(f"\nFiles saved to: {output_dir}")


if __name__ == "__main__":
    main()

File: MATLAB/run_optimization.m
function output_dir = run_optimization(config_path)
% RUN_OPTIMIZATION  Top-level pipeline: load config, optimize, save results.
%
%   output_dir = RUN_OPTIMIZATION(config_path)
%
%   MATLAB port of scripts/run_optimization.py. Loads a YAML config, parses CST
%   element patterns, runs the multi-start fmincon optimizer, evaluates metrics,
%   saves plots (+ optional GIF), and writes weights.csv / metrics.json /
%   run_report.txt / a config snapshot to a timestamped results subfolder.
%
%   Inputs:
%       config_path : path to config.yaml. Default: <repo>/config.yaml.
%
%   Outputs:
%       output_dir : the timestamped results directory that was created.
%
%   Part of: Antenna Array Pattern Optimization Tool.

REPO_ROOT = fileparts(fileparts(mfilename('fullpath')));
if nargin < 1 || isempty(config_path)
    config_path = fullfile(REPO_ROOT, 'config.yaml');
end
% A relative config name not found in the current folder is resolved against the
% repo root, so the script works regardless of the working directory.
if ~isfile(config_path) && isfile(fullfile(REPO_ROOT, config_path))
    config_path = fullfile(REPO_ROOT, config_path);
end

% Special polarization keyword: incoherent power sum across all detected
% polarization components (e.g. |AF_Copol|^2 + |AF_Cross|^2, or
% |AF_Theta|^2 + |AF_Phi|^2). Only supported when exactly two components are
% present, since the optimizer/cost function take a single secondary stack.
POLARIZATION_TOTAL = 'total';

% Default polarization when not specified in config.yaml.
DEFAULT_POLARIZATION = 'copol';

REQUIRED_CONFIG_KEYS = {'element_patterns_dir', 'directives', 'optimizer', 'output'};

% ── Load and validate config ──────────────────────────────────────
config = read_config_yaml(config_path);
for i = 1:numel(REQUIRED_CONFIG_KEYS)
    if ~isfield(config, REQUIRED_CONFIG_KEYS{i})
        error('run_optimization:MissingKey', ...
            'Missing required config key: ''%s''.', REQUIRED_CONFIG_KEYS{i});
    end
end
polarization = get_field(config, 'polarization', DEFAULT_POLARIZATION);

% ── Timestamped output directory ──────────────────────────────────
results_dir = resolve_path(config.output.results_dir, REPO_ROOT);
timestamp   = char(datetime('now', 'Format', 'yyyy-MM-dd_HHmmss'));
output_dir  = fullfile(results_dir, 'optimizations', timestamp);
if ~isfolder(output_dir), mkdir(output_dir); end
fprintf('Output directory: %s\n', output_dir);

% ── Stage 1: load element patterns ────────────────────────────────
fprintf('Loading element patterns...\n');
patterns_dir = resolve_path(config.element_patterns_dir, REPO_ROOT);
patterns  = load_element_patterns(patterns_dir);
theta_deg = patterns(1).theta_deg;
phi_deg   = patterns(1).phi_deg;
fprintf('  Polarization: %s\n', polarization);

% Component names are matched case-insensitively against whatever
% Abs(<name>)/Phase(<name>) pairs were detected in the CST export header
% (e.g. 'Copol'/'Cross' or 'Theta'/'Phi').
component_names = sort(fieldnames(patterns(1).components));

if strcmpi(polarization, POLARIZATION_TOTAL)
    % Total field: incoherent power sum across all detected components,
    % e.g. |AF_Copol|^2 + |AF_Cross|^2 or |AF_Theta|^2 + |AF_Phi|^2.
    if numel(component_names) ~= 2
        error('run_optimization:BadPolarization', ...
            'polarization: ''total'' requires exactly 2 polarization components, found {%s}.', ...
            strjoin(component_names, ', '));
    end
    name_primary   = component_names{1};
    name_secondary = component_names{2};
    element_patterns_stacked   = stack_component(patterns, name_primary);
    element_patterns_secondary = stack_component(patterns, name_secondary);
    fprintf('  Using ''%s'' + ''%s'' -> total power sum.\n', name_primary, name_secondary);
else
    matches = component_names(strcmpi(component_names, polarization));
    if isempty(matches)
        error('run_optimization:BadPolarization', ...
            'Polarization ''%s'' not found. Available components: {%s} (or ''total'').', ...
            polarization, strjoin(component_names, ', '));
    end
    fprintf('  Using component ''%s''\n', matches{1});
    element_patterns_stacked   = stack_component(patterns, matches{1});
    element_patterns_secondary = [];
end
[n_elements, n_theta, n_phi] = size(element_patterns_stacked);
fprintf('  Loaded %d elements, grid %dx%d\n', n_elements, n_theta, n_phi);

% ── Stage 4: optimizer ────────────────────────────────────────────
fprintf('Running optimizer...\n');
directives = config.directives;
t_start    = tic;
opt = run_optimizer(element_patterns_stacked, theta_deg, phi_deg, ...
    directives, config.optimizer, element_patterns_secondary);
elapsed_sec = toc(t_start);

weights_complex = opt.weights_complex;
cost_history    = opt.cost_history;

% ── Power grid (metrics + plotting) ───────────────────────────────
af_primary = compute_array_factor(weights_complex, element_patterns_stacked);
if ~isempty(element_patterns_secondary)
    af_secondary = compute_array_factor(weights_complex, element_patterns_secondary);
    power_linear = abs(af_primary) .^ 2 + abs(af_secondary) .^ 2;
    metrics_af   = sqrt(power_linear);   % synthetic real AF: |sqrt(P)|^2 == P
else
    power_linear = abs(af_primary) .^ 2;
    metrics_af   = [];
end
array_factor_db_grid = 10 * log10(max(power_linear, 1e-30));

% ── Stage 5: metrics ──────────────────────────────────────────────
metrics = evaluate_metrics(element_patterns_stacked, theta_deg, phi_deg, ...
    weights_complex, directives, cost_history, metrics_af);

% ── Stage 6: plots ────────────────────────────────────────────────
fprintf('Saving plots...\n');
save_all_plots(array_factor_db_grid, theta_deg, phi_deg, weights_complex, directives, ...
    opt.all_cost_histories, opt.best_run_index, opt.all_run_labels, config.output, output_dir);

% ── Stage 7: pattern GIF (optional) ──────────────────────────────
if get_field(config.output, 'save_pattern_gif', false)
    fprintf('Saving pattern evolution GIF...\n');
    save_pattern_gif(opt.weights_history, element_patterns_stacked, theta_deg, phi_deg, ...
        directives, cost_history, config.output, output_dir, element_patterns_secondary);
end

% ── Save weights / metrics / report / config snapshot ────────────
save_weights_csv(weights_complex, fullfile(output_dir, 'weights.csv'));
save_metrics_json(metrics, fullfile(output_dir, 'metrics.json'));
save_run_report(metrics, opt, elapsed_sec, config, fullfile(output_dir, 'run_report.txt'));
copyfile(config_path, fullfile(output_dir, 'config.yaml'));

% ── Summary ───────────────────────────────────────────────────────
fprintf('\nResults:\n');
fprintf('  Optimization time:  %.1f s\n', elapsed_sec);
fprintf('  Converged:          %d\n', opt.result.success);
fprintf('  Iterations:         %d\n', metrics.iteration_count);
fprintf('  Final cost J:       %.6f\n', metrics.total_cost);
fprintf('  Global peak:        %.2f dBi\n', metrics.global_peak_dbi);
fprintf('\nFiles saved to: %s\n', output_dir);
end


% ────────────────────────── HELPERS ───────────────────────────────

function val = get_field(s, name, default)
if isfield(s, name) && ~isempty(s.(name))
    val = s.(name);
else
    val = default;
end
end


function p = resolve_path(p, repo_root)
% Resolve a possibly-relative config path against the repo root. Absolute or
% already-existing paths are returned unchanged (do NOT prepend repo_root).
if isfolder(p) || isfile(p) || is_absolute_path(p)
    return;
end
p = fullfile(repo_root, p);
end


function tf = is_absolute_path(p)
% True for a Windows drive path (C:\...), UNC (\\...), or POSIX-rooted (/...).
tf = ~isempty(regexp(p, '^([A-Za-z]:[\\/]|[\\/])', 'once'));
end


function save_weights_csv(weights_complex, output_path)
% Columns: element_index, amplitude, phase_deg, real, imag (2-decimal like Python).
fid = fopen(output_path, 'w');
fprintf(fid, 'element_index,amplitude,phase_deg,real,imag\n');
for n = 1:numel(weights_complex)
    w = weights_complex(n);
    fprintf(fid, '%d,%.2f,%.2f,%.2f,%.2f\n', ...
        n - 1, abs(w), angle(w) * 180 / pi, real(w), imag(w));
end
fclose(fid);
end


function save_metrics_json(metrics, output_path)
% Convert empty ([] = Python None) to NaN so jsonencode emits null.
m = metrics;
for k = 1:numel(m.directive_metrics)
    if isempty(m.directive_metrics(k).null_depth_db)
        m.directive_metrics(k).null_depth_db = NaN;
    end
end
if isempty(m.peak_to_null_ratio_db)
    m.peak_to_null_ratio_db = NaN;
end
fid = fopen(output_path, 'w');
fprintf(fid, '%s', jsonencode(m, 'PrettyPrint', true));
fclose(fid);
end


function save_run_report(metrics, opt, elapsed_sec, config, output_path)
opt_cfg    = config.optimizer;
directives = config.directives;
lines = {};
bar = repmat('=', 1, 56);
lines{end+1} = bar;
lines{end+1} = 'Antenna Array Optimization Report';
lines{end+1} = char(datetime('now', 'Format', 'yyyy-MM-dd HH:mm:ss'));
lines{end+1} = bar;
lines{end+1} = '';
lines{end+1} = 'RUN INFO';
lines{end+1} = sprintf('  Elapsed time:      %.1f s', elapsed_sec);
lines{end+1} = sprintf('  Converged:         %d', opt.result.success);
lines{end+1} = sprintf('  Iterations:        %d', metrics.iteration_count);
lines{end+1} = sprintf('  Final cost J:      %.6f', metrics.total_cost);
lines{end+1} = '';
lines{end+1} = 'OPTIMIZER SETTINGS';
lines{end+1} = '  Method:            fmincon (interior-point)';
lines{end+1} = sprintf('  Max iterations:    %s', num2str(get_field(opt_cfg, 'max_iterations', NaN)));
lines{end+1} = sprintf('  Cost tolerance:    %g', get_field(opt_cfg, 'cost_tolerance', NaN));
lines{end+1} = sprintf('  N restarts:        %s', num2str(get_field(opt_cfg, 'n_restarts', NaN)));
lines{end+1} = '';
lines{end+1} = 'DIRECTIVES';
for i = 1:numel(directives)
    d = directives{i};
    sym = get_field(d, 'width', NaN);
    lines{end+1} = sprintf('  [%d] %-4s  theta=%.1f  phi=%.1f  theta_width=%s  phi_width=%s  weight=%g', ...
        i - 1, d.type, d.theta, get_field(d, 'phi', 0.0), ...
        num2str(get_field(d, 'theta_width', sym)), num2str(get_field(d, 'phi_width', sym)), ...
        get_field(d, 'weight', 1.0));
end
lines{end+1} = '';
lines{end+1} = 'RESULTS';
lines{end+1} = sprintf('  Global peak:       %.2f dBi', metrics.global_peak_dbi);
if ~isempty(metrics.peak_to_null_ratio_db) && ~isnan(metrics.peak_to_null_ratio_db)
    lines{end+1} = sprintf('  Peak-to-null:      %.2f dB', metrics.peak_to_null_ratio_db);
end
lines{end+1} = '';
for i = 1:numel(metrics.directive_metrics)
    dm = metrics.directive_metrics(i);
    if strcmp(dm.type, 'peak')
        lines{end+1} = sprintf('  Peak @ theta=%.1f, phi=%.1f:  %.2f dBi', ...
            dm.theta_deg, dm.phi_deg, dm.gain_dbi);
    else
        lines{end+1} = sprintf('  Null @ theta=%.1f, phi=%.1f:  %.2f dBi  (depth = %.2f dB)', ...
            dm.theta_deg, dm.phi_deg, dm.gain_dbi, dm.null_depth_db);
    end
end
lines{end+1} = '';
lines{end+1} = bar;

fid = fopen(output_path, 'w');
fprintf(fid, '%s\n', lines{:});
fclose(fid);
end
